consensus/src/chained_bft/block_storage/block_tree.rs

```diff
@@ -22,13 +19,56 @@ use logger::prelude::*;
 use mirai_annotations::{checked_verify_eq, precondition};
 use serde::Serialize;
 use std::{
-    collections::{vec_deque::VecDeque, HashMap},
+    collections::{vec_deque::VecDeque, HashMap, HashSet},
     fmt::Debug,
+    ops::Deref,
     sync::Arc,
     time::Duration,
 };
 use types::crypto_proxies::LedgerInfoWithSignatures;
 
+/// This structure is a wrapper of [`ExecutedBlock`](crate::consensus_types::block::ExecutedBlock)
+/// that adds `children` field to know the parent-child relationship between blocks.
+struct LinkableBlock<T> {
+    /// Executed block that has raw block data and execution output.
+    executed_block: Arc<ExecutedBlock<T>>,
+    /// The set of children for cascading pruning. Note: a block may have multiple children.
+    children: HashSet<HashValue>,
```

Comment:
can this be a `List<Arc<ExecutedBlock>>`?

Reply:
In `BlockTree` we want get child as a `LinkableBlock` but not an `ExecutedBlock`. So it is reasonable to change it to `Vec<&LinkableBlock>`, but it will make the children blocks immutable because of the reference. That's why I propose to use hash everywhere instead of `Arc<T>`.